### DISCUSSION TITLE
Anyone have experience auto-generating external scanner lexing code?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1322] [Q&A] [answered]
My external scanner lexing code has grown to be more and more gnarly over time, having reached about 350 lines of code like the following (bearing some resemblance to tree-sitter's auto-generated lexing code):
```cpp
...
 case LexState::VARIABLE:
  ACCEPT_LEXEME(Lexeme::VARIABLE_KEYWORD);
  if ('S' == lookahead) ADVANCE(LexState::VARIABLES);
  if (is_identifier_char(lookahead)) ADVANCE(LexState::IDENTIFIER);
  END_LEX_STATE();
case LexState::VARIABLES:
  ACCEPT_LEXEME(Lexeme::VARIABLES_KEYWORD);
  if (is_identifier_char(lookahead)) ADVANCE(LexState::IDENTIFIER);
  END_LEX_STATE();
...
```
I appreciate the flexibility this low-level lexing code provides but am interested in whether anyone has had success auto-generating this stuff instead. I have to lookahead for a lot of keywords that could terminate an indented block (regardless of their own indentation), among other things.

**Comment by ahlinc** [comment, marked as answer]
I remember that @alemuller mentioned [re2c](https://re2c.org/) in https://github.com/tree-sitter/tree-sitter/discussions/1092#discussioncomment-714008 and there were [other](https://github.com/tree-sitter/tree-sitter/issues/885#issuecomment-766326397) [mentions](https://github.com/tree-sitter/tree-sitter/issues/621#issuecomment-629771852) too.
I also had thoughts about the same but my concern is that having another one lexer generator would make debugging even more difficult.
_And your `scanner.cc` for the TLA+ grammar looks pretty nice :+1:_

**Reply by ahelwer** [reply]
Thanks!
